core/src/main/scala/kafka/utils/CoreUtils.scala

```diff
@@ -303,4 +303,13 @@ object CoreUtils extends Logging {
     props.foreach { case (k, v) => properties.put(k, v) }
     properties
   }
+
+  def getOrElseUpdateAtomically[K, V](map: concurrent.Map[K, V], key: K, createValue: => V): V = {
```

Comment:
Nit: I think we should call this `atomicGetOrUpdate`, but we should go ahead and merge if Rajini is already offline.

Reply:
Also, we should document the method and state that it's optimised for the case where keys exist more times than they're created.